File: Calto-GG/Microsoft/Day 3/CourseSchedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        from collections import defaultdict
        graph = defaultdict(list)
        for a, b in prerequisites:
            graph[b].append(a)

        # Create an array to track the in-degree (number of incoming edges) of each node
        in_degree = [0 for _ in range(numCourses)]
        for node in graph:
            for neighbor in graph[node]:
                in_degree[neighbor] += 1

        # Initialize a queue to hold nodes with no incoming edges (i.e. no prerequisites)
        queue = []
        for i in range(numCourses):
            if in_degree[i] == 0:
                queue.append(i)

        # Perform a topological sort
        while queue:
            node = queue.pop(0)
            for neighbor in graph[node]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        # If all nodes have been visited, a valid ordering exists and we return True
        # Otherwise, we return False
        return sum(in_degree) == 0

File: Harishri/Microsoft/Day 3/Course_Schedule.cpp
class Solution {
	public:
		bool canFinish(int n, vector<vector<int>>& pre) {
			vector<int> adj[n];
			for(int i=0;i<pre.size();i++) adj[pre[i][1]].push_back(pre[i][0]);
			vector<int> topo;
			vector<int> in(n,0);
			queue<int>q;
			for(int i=0;i<n;i++){
				for(auto j:adj[i]) in[j]++;
			}
			for(int node=0;node<n;node++){
				if(!in[node]) q.push(node);
			}
			while(!q.empty()){
				int node=q.front();
				q.pop();
				topo.push_back(node);
				for(auto i:adj[node]){
					in[i]--;
					if(!in[i]) q.push(i);
				}
			}
			return topo.size()==n;
		}
	};
